relay/cached_accessor.go

```diff
@@ -0,0 +1,118 @@
+package relay
+
+import (
+	lru "github.com/hashicorp/golang-lru/v2"
+	"sync"
+)
+
+// CachedAccessor is an interface for accessing a resource that is cached. It assumes that cache misses
+// are expensive, and prevents multiple concurrent cache misses for the same key.
+type CachedAccessor[K comparable, V any] interface {
```

Comment:
Does it make sense to not generalize this interface for now as the only use case is retrieving blobs from the store? I'm generally against generalizing something like this unless there are 3+ immediate use cases.
I think the interface will be a lot simpler (and don't need to pass in function as `Accessor`) if we just made this `CachedBlobStore` or something that has the s3 store and accessor logic implemented directly. Wdyt?

Reply:
I agree with your general sentiment on this, making something generic adds complexity to its implementation. But in this case, there are actually three types of data that I'd like to use this pattern for:

- fetching blob metadata
- fetching unencoded blobs
- fetching encoded chunks

I expect that preventing parallel fetches for each of these data types in the advent of a cache miss will be beneficial.

Reply:
I see. Makes sense